src/server/conn.rs

```diff
@@ -513,3 +553,165 @@ where
         }
     }
 }
+
+// TODO: different file? idk where to put this.
+mod rewindable {
+    use bytes::{Buf, BufMut, Bytes, IntoBuf};
+    use futures::{Async, Poll};
+    use std::io::{self, Read, Write};
+    use tokio_io::{AsyncRead, AsyncWrite};
+
+    #[derive(Debug)]
+    pub struct Rewindable<T>
+    where
+        T: AsyncRead + AsyncWrite,
```

Comment:
Ah, I see. If these bounds are removed, then they won't be needed in several other places as well. Usually, prefer bounds only on `impl`s that need them (sometimes it's unavoidable in a struct definition, if you need to refer to an associated type).

Reply:
This is a great rust nugget of knowledge, thanks.